internal/testutil/acc/config_formatter.go

```diff
@@ -50,3 +57,184 @@ func sortStringMapKeys(m map[string]string) []string {
+
+func ToSnakeCase(str string) string {
+	snake := matchFirstCap.ReplaceAllString(str, "${1}_${2}")
+	snake = matchAllCap.ReplaceAllString(snake, "${1}_${2}")
+	return strings.ToLower(snake)
+}
+
+func ClusterResourceHcl(projectID string, req *ClusterRequest) (configStr, clusterName string, err error) {
+	if req == nil {
+		req = new(ClusterRequest)
+	}
+	specs := req.ReplicationSpecs
+	if len(specs) == 0 {
+		specs = append(specs, ReplicationSpec(&ReplicationSpecRequest{}))
+	}
+	clusterName = req.ClusterNameExplicit
+	if clusterName == "" {
+		clusterName = RandomClusterName()
+	}
+	clusterTypeStr := "REPLICASET"
+	if req.Geosharded {
+		clusterTypeStr = "GEOSHARDED"
+	}
+
+	f := hclwrite.NewEmptyFile()
+	root := f.Body()
+	cluster := root.AppendNewBlock("resource", []string{"mongodbatlas_advanced_cluster", "cluster_info"}).Body()
+	clusterRootAttributes := map[string]any{
+		"project_id":     projectID,
+		"cluster_type":   clusterTypeStr,
+		"name":           clusterName,
+		"backup_enabled": req.CloudBackup,
+	}
+	if req.DiskSizeGb != 0 {
+		clusterRootAttributes["disk_size_gb"] = req.DiskSizeGb
+	}
+	addPrimitiveAttributes(cluster, clusterRootAttributes)
+	cluster.AppendNewline()
+	for i, spec := range specs {
+		err = writeReplicationSpec(cluster, spec)
+		if err != nil {
+			return "", "", fmt.Errorf("error writing hcl for replication spec %d: %w", i, err)
+		}
+	}
+	cluster.AppendNewline()
+	if req.ResourceDependencyName != "" {
+		if !strings.Contains(req.ResourceDependencyName, ".") {
+			return "", "", fmt.Errorf("req.ResourceDependencyName must have a '.'")
+		}
+		err = setAttributeHcl(cluster, fmt.Sprintf("depends_on = [%s]", req.ResourceDependencyName))
+		if err != nil {
+			return "", "", err
+		}
+	}
+	return "\n" + string(f.Bytes()), clusterName, err
+}
+
+func writeReplicationSpec(cluster *hclwrite.Body, spec admin.ReplicationSpec) error {
+	replicationBlock := cluster.AppendNewBlock("replication_specs", nil).Body()
+	err := addPrimitiveAttributesViaJSON(replicationBlock, spec)
+	if err != nil {
+		return err
+	}
+	for _, rc := range spec.GetRegionConfigs() {
+		if rc.Priority == nil {
+			rc.SetPriority(7)
+		}
+		replicationBlock.AppendNewline()
+		rcBlock := replicationBlock.AppendNewBlock("region_configs", nil).Body()
+		err = addPrimitiveAttributesViaJSON(rcBlock, rc)
+		if err != nil {
+			return err
+		}
+		autoScalingBlock := rcBlock.AppendNewBlock("auto_scaling", nil).Body()
+		if rc.AutoScaling == nil {
+			autoScalingBlock.SetAttributeValue("disk_gb_enabled", cty.BoolVal(false))
+		} else {
+			autoScaling := rc.GetAutoScaling()
+			return fmt.Errorf("auto_scaling on replication spec is not supportd yet %v", autoScaling)
+		}
+		nodeSpec := rc.GetElectableSpecs()
+		nodeSpecBlock := rcBlock.AppendNewBlock("electable_specs", nil).Body()
+		err = addPrimitiveAttributesViaJSON(nodeSpecBlock, nodeSpec)
+	}
+	return err
+}
+
+// Helper function for adding "primitive" bool/string/int/float attributes of a struct.
```

Comment:
```suggestion
// addPrimitiveAttributesViaJSON  adds "primitive" bool/string/int/float attributes of a struct.
```

function comments should start with the function name

Reply:
changed in bfb2a5e7939df716a73f6c3d4ee661214082a21f